File: backend/services/excel_data_service.py
"""
Excel Data Service
Handles reading, filtering, searching, and pagination of Excel files
"""

import pandas as pd
import os
from typing import Dict, List, Any, Optional, Tuple
import math
from constants.file_constants import DATABASE_DIR

class ExcelDataService:
    """Service for handling Excel file data operations"""
    
    AVAILABLE_FILES = {
        'vehicle_fleet_master_data': 'vehicle_fleet_master_data.xlsx',
        'equipment_lifecycle_reference': 'equipment_lifecycle_reference.xlsx',
        'equipment_lifecycle_by_business_unit': 'equipment_lifecycle_by_business_unit.xlsx',
        'vehicle_replacement_by_category': 'vehicle_replacement_by_category.xlsx',
        'vehicle_replacement_detailed_forecast': 'vehicle_replacement_detailed_forecast.xlsx',
        'radio_equipment_cost_analysis': 'radio_equipment_cost_analysis.xlsx',
        'electric_vehicle_budget_analysis': 'electric_vehicle_budget_analysis.xlsx',
        'user': 'user.xlsx'
    }
    
    @staticmethod
    def get_available_files() -> Dict[str, str]:
        """Get list of available Excel files"""
        available = {}
        for key, filename in ExcelDataService.AVAILABLE_FILES.items():
            file_path = os.path.join(DATABASE_DIR, filename)
            if os.path.exists(file_path):
                available[key] = filename
        return available
    
    @staticmethod
    def read_excel_file(file_key: str, sheet_name: Optional[str] = None) -> pd.DataFrame:
        """Read Excel file and return DataFrame"""
        if file_key not in ExcelDataService.AVAILABLE_FILES:
            raise ValueError(f"File '{file_key}' not found in available files")
        
        filename = ExcelDataService.AVAILABLE_FILES[file_key]
        file_path = os.path.join(DATABASE_DIR, filename)
        
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Excel file not found: {file_path}")
        
        try:
            # Read Excel file
            if sheet_name:
                df = pd.read_excel(file_path, sheet_name=sheet_name)
            else:
                df = pd.read_excel(file_path)
            
            # Clean column names
            df.columns = df.columns.astype(str)
            
            # Fill NaN values with empty strings for better JSON serialization
            df = df.fillna('')
            
            return df
        except Exception as e:
            raise Exception(f"Error reading Excel file: {str(e)}")
    
    @staticmethod
    def get_file_info(file_key: str) -> Dict[str, Any]:
        """Get information about an Excel file"""
        try:
            df = ExcelDataService.read_excel_file(file_key)
            
            # Get Excel file sheets
            filename = ExcelDataService.AVAILABLE_FILES[file_key]
            file_path = os.path.join(DATABASE_DIR, filename)
            
            try:
                excel_file = pd.ExcelFile(file_path)
                sheets = excel_file.sheet_names
            except:
                sheets = ['Sheet1']  # Default fallback
            
            return {
                'file_key': file_key,
                'filename': filename,
                'total_rows': len(df),
                'total_columns': len(df.columns),
                'columns': list(df.columns),
                'sheets': sheets,
                'file_path': file_path,
                'file_exists': True
            }
        except Exception as e:
            return {
                'file_key': file_key,
                'filename': ExcelDataService.AVAILABLE_FILES.get(file_key, ''),
                'error': str(e),
                'file_exists': False
            }
    
    @staticmethod
    def filter_and_search_data(
        df: pd.DataFrame,
        search_query: Optional[str] = None,
        column_filters: Optional[Dict[str, Any]] = None,
        value_filters: Optional[Dict[str, Any]] = None,
        sort_column: Optional[str] = None,
        sort_direction: str = 'asc'
    ) -> pd.DataFrame:
        """Apply filters, search, and sorting to DataFrame"""
        
        filtered_df = df.copy()
        
        # Apply column-specific filters
        if column_filters:
            for column, filter_value in column_filters.items():
                if column in filtered_df.columns and filter_value:
                    if isinstance(filter_value, str):
                        # String contains filter (case-insensitive)
                        filtered_df = filtered_df[
                            filtered_df[column].astype(str).str.contains(
                                filter_value, case=False, na=False
                            )
                        ]
                    elif isinstance(filter_value, (int, float)):
                        # Exact numeric match
                        filtered_df = filtered_df[filtered_df[column] == filter_value]
                    elif isinstance(filter_value, list):
                        # Multiple values filter
                        filtered_df = filtered_df[filtered_df[column].isin(filter_value)]
        
        # Apply value filters (range filters for numeric columns)
        if value_filters:
            for column, filter_config in value_filters.items():
                if column in filtered_df.columns:
                    if isinstance(filter_config, dict):
                        min_val = filter_config.get('min')
                        max_val = filter_config.get('max')
                        
                        if min_val is not None:
                            filtered_df = filtered_df[
                                pd.to_numeric(filtered_df[column], errors='coerce') >= min_val
                            ]
                        if max_val is not None:
                            filtered_df = filtered_df[
                                pd.to_numeric(filtered_df[column], errors='coerce') <= max_val
                            ]
        
        # Apply global search query
        if search_query:
            search_mask = pd.Series([False] * len(filtered_df))
            
            for column in filtered_df.columns:
                column_mask = filtered_df[column].astype(str).str.contains(
                    search_query, case=False, na=False
                )
                search_mask = search_mask | column_mask
            
            filtered_df = filtered_df[search_mask]
        
        # Apply sorting
        if sort_column and sort_column in filtered_df.columns:
            ascending = sort_direction.lower() == 'asc'
            filtered_df = filtered_df.sort_values(
                by=sort_column, 
                ascending=ascending,
                na_position='last'
            )
        
        return filtered_df
    
    @staticmethod
    def paginate_data(
        df: pd.DataFrame,
        page: int = 1,
        page_size: int = 50
    ) -> Tuple[pd.DataFrame, Dict[str, Any]]:
        """Apply pagination to DataFrame and return data with metadata"""
        
        total_rows = len(df)
        total_pages = math.ceil(total_rows / page_size) if page_size > 0 else 1
        
        # Ensure page is within bounds
        page = max(1, min(page, total_pages)) if total_pages > 0 else 1
        
        # Calculate start and end indices
        start_idx = (page - 1) * page_size
        end_idx = start_idx + page_size
        
        # Get paginated data
        paginated_df = df.iloc[start_idx:end_idx]
        
        # Create pagination metadata
        pagination_info = {
            'current_page': page,
            'page_size': page_size,
            'total_rows': total_rows,
            'total_pages': total_pages,
            'has_next': page < total_pages,
            'has_previous': page > 1,
            'start_row': start_idx + 1 if total_rows > 0 else 0,
            'end_row': min(end_idx, total_rows)
        }
        
        return paginated_df, pagination_info
    
    @staticmethod
    def get_column_stats(df: pd.DataFrame, column: str) -> Dict[str, Any]:
        """Get statistics for a specific column"""
        if column not in df.columns:
            return {'error': f"Column '{column}' not found"}
        
        col_data = df[column]
        stats = {
            'column_name': column,
            'total_values': len(col_data),
            'non_null_values': col_data.count(),
            'null_values': col_data.isnull().sum(),
            'unique_values': col_data.nunique(),
            'data_type': str(col_data.dtype)
        }
        
        # Add numeric statistics if applicable
        numeric_data = pd.to_numeric(col_data, errors='coerce')
        if not numeric_data.isnull().all():
            stats.update({
                'min_value': numeric_data.min(),
                'max_value': numeric_data.max(),
                'mean_value': numeric_data.mean(),
                'median_value': numeric_data.median(),
                'std_deviation': numeric_data.std()
            })
        
        # Add top values
        try:
            value_counts = col_data.value_counts().head(10)
            stats['top_values'] = [
                {'value': str(val), 'count': int(count)} 
                for val, count in value_counts.items()
            ]
        except:
            stats['top_values'] = []
        
        return stats
    
    @staticmethod
    def get_data_with_filters(
        file_key: str,
        page: int = 1,
        page_size: int = 50,
        search_query: Optional[str] = None,
        column_filters: Optional[Dict[str, Any]] = None,
        value_filters: Optional[Dict[str, Any]] = None,
        sort_column: Optional[str] = None,
        sort_direction: str = 'asc',
        sheet_name: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get filtered and paginated data from Excel file"""
        
        try:
            # Read the Excel file
            df = ExcelDataService.read_excel_file(file_key, sheet_name)
            
            # Apply filters and search
            filtered_df = ExcelDataService.filter_and_search_data(
                df=df,
                search_query=search_query,
                column_filters=column_filters,
                value_filters=value_filters,
                sort_column=sort_column,
                sort_direction=sort_direction
            )
            
            # Apply pagination
            paginated_df, pagination_info = ExcelDataService.paginate_data(
                filtered_df, page, page_size
            )
            
            # Convert to records for JSON serialization
            data = paginated_df.to_dict('records')
            
            return {
                'success': True,
                'data': data,
                'columns': list(df.columns),
                'pagination': pagination_info,
                'filters_applied': {
                    'search_query': search_query,
                    'column_filters': column_filters,
                    'value_filters': value_filters,
                    'sort_column': sort_column,
                    'sort_direction': sort_direction
                },
                'file_info': {
                    'file_key': file_key,
                    'sheet_name': sheet_name,
                    'total_rows_before_filter': len(df),
                    'total_rows_after_filter': len(filtered_df)
                }
            }
            
        except Exception as e:
            return {
                'success': False,
                'error': str(e),
                'file_key': file_key
            }

File: frontend/src/services/excelDataService.ts
const API_BASE_URL = 'http://localhost:3300/api/excel';

export interface ExcelColumnStats {
  column_name: string;
  total_values: number;
  non_null_values: number;
  null_values: number;
  unique_values: number;
  data_type: string;
  min_value?: number;
  max_value?: number;
  mean_value?: number;
  median_value?: number;
  std_deviation?: number;
  top_values: Array<{value: string; count: number}>;
}

export interface ExcelFileInfo {
  file_key: string;
  filename: string;
  total_rows: number;
  total_columns: number;
  columns: string[];
  sheets: string[];
  file_exists: boolean;
  error?: string;
}

export interface ExcelPagination {
  current_page: number;
  page_size: number;
  total_rows: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
  start_row: number;
  end_row: number;
}

export interface ExcelDataFilters {
  search_query?: string;
  column_filters?: Record<string, any>;
  value_filters?: Record<string, {min?: number; max?: number}>;
  sort_column?: string;
  sort_direction?: 'asc' | 'desc';
}

export interface ExcelDataResponse {
  success: boolean;
  data: Record<string, any>[];
  columns: string[];
  pagination: ExcelPagination;
  filters_applied: ExcelDataFilters;
  file_info: {
    file_key: string;
    sheet_name?: string;
    total_rows_before_filter: number;
    total_rows_after_filter: number;
  };
  error?: string;
}

export interface ExcelDataRequest {
  page?: number;
  page_size?: number;
  search?: string;
  column_filters?: Record<string, any>;
  value_filters?: Record<string, {min?: number; max?: number}>;
  sort_column?: string;
  sort_direction?: 'asc' | 'desc';
  sheet?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

class ExcelDataService {
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    try {
      console.log('Excel API Request:', `${API_BASE_URL}${endpoint}`);
      
      const response = await fetch(`${API_BASE_URL}${endpoint}`, {
        headers: {
          'Content-Type': 'application/json',
          ...options.headers,
        },
        ...options,
      });

      console.log('Excel API Response Status:', response.status);
      
      const result = await response.json();
      console.log('Excel API Response Data:', result);
      
      if (!response.ok) {
        console.error('Excel API Error:', result);
        return {
          success: false,
          error: result.error || result.message || `HTTP ${response.status}`,
        };
      }

      // Excel API returns the full structure: {success, data, columns, pagination, file_info, filters_applied}
      // We need to return just the structured part (without the outer success wrapper)
      if (result.success !== undefined) {
        // Extract everything except the success field for the data
        const { success, ...dataStructure } = result;
        return {
          success: result.success,
          data: dataStructure as T,
          error: result.error
        };
      } else {
        // Fallback for other formats
        return {
          success: true,
          data: result as T,
        };
      }
    } catch (error) {
      console.error('Excel API Network Error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Network error',
      };
    }
  }

  // Get available Excel files
  async getAvailableFiles() {
    return this.request<{
      available_files: Record<string, string>;
      file_details: Record<string, ExcelFileInfo>;
    }>('/files');
  }

  // Get file information
  async getFileInfo(fileKey: string) {
    return this.request<{ file_info: ExcelFileInfo }>(`/files/${fileKey}/info`);
  }

  // Get file data (GET method with query params)
  async getFileData(fileKey: string, params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    const searchParams = new URLSearchParams();
    
    if (params.page) searchParams.set('page', params.page.toString());
    if (params.page_size) searchParams.set('page_size', params.page_size.toString());
    if (params.search) searchParams.set('search', params.search);
    if (params.sort_column) searchParams.set('sort_column', params.sort_column);
    if (params.sort_direction) searchParams.set('sort_direction', params.sort_direction);
    if (params.sheet) searchParams.set('sheet', params.sheet);
    
    // Add column filters
    if (params.column_filters) {
      Object.entries(params.column_filters).forEach(([key, value]) => {
        if (value && value !== 'all') {
          searchParams.set(`filter_${key}`, Array.isArray(value) ? value.join(',') : value);
        }
      });
    }
    
    // Add value filters as JSON
    if (params.value_filters) {
      searchParams.set('value_filters', JSON.stringify(params.value_filters));
    }

    const queryString = searchParams.toString();
    const endpoint = `/files/${fileKey}/data${queryString ? `?${queryString}` : ''}`;
    
    return this.request<ExcelDataResponse>(endpoint);
  }

  // Get file data (POST method for complex filters)
  async getFileDataPost(fileKey: string, data: ExcelDataRequest): Promise<ApiResponse<ExcelDataResponse>> {
    return this.request<ExcelDataResponse>(`/files/${fileKey}/data`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // Get column statistics
  async getColumnStats(fileKey: string, column: string, sheet?: string) {
    const searchParams = new URLSearchParams();
    if (sheet) searchParams.set('sheet', sheet);
    
    const endpoint = `/files/${fileKey}/columns/${encodeURIComponent(column)}/stats${
      searchParams.toString() ? `?${searchParams.toString()}` : ''
    }`;
    
    return this.request<{ column_stats: ExcelColumnStats }>(endpoint);
  }

  // Global search across all files
  async globalSearch(query: string, maxResults: number = 10) {
    const searchParams = new URLSearchParams({
      search: query,
      max_results: maxResults.toString(),
    });
    
    return this.request<{
      search_query: string;
      results: Record<string, {
        matches_found: number;
        total_matches: number;
        sample_data: Record<string, any>[];
        columns: string[];
      }>;
      files_searched: string[];
    }>(`/search?${searchParams.toString()}`);
  }

  // Shortcut methods for commonly used files
  async getVehicleFleetData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    console.log('Getting vehicle fleet data with params:', params);
    return this.getFileData('vehicle_fleet_master_data', params);
  }

  async getEquipmentLifecycleData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('equipment_lifecycle_reference', params);
  }

  async getElectricVehicleBudgetData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('electric_vehicle_budget_analysis', params);
  }

  async getRadioEquipmentCostData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('radio_equipment_cost_analysis', params);
  }

  async getVehicleReplacementForecastData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('vehicle_replacement_detailed_forecast', params);
  }

  async getVehicleReplacementCategoryData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('vehicle_replacement_by_category', params);
  }

  async getEquipmentLifecycleBusinessData(params: ExcelDataRequest = {}): Promise<ApiResponse<ExcelDataResponse>> {
    return this.getFileData('equipment_lifecycle_by_business_unit', params);
  }
}

export const excelDataService = new ExcelDataService();